### PR TITLE
Adds an overwrite flag to force annotation updates

### PR DESCRIPTION
**Pull request by tony-landreth**
Issue: PGO-39

### CODE REVIEW

File: internal/cmd/backup.go [hunk 1]
```diff
@@ -107,16 +108,21 @@ pgo backup hippo --repoName="repo1" --options="--type=full"
...
 			args[0], // the name of the cluster object, limited to one name through `ExactArgs(1)`
 			types.ApplyPatchType,
 			patch,
-			config.Patch.PatchOptions(metav1.PatchOptions{}),
+			config.Patch.PatchOptions(patchOptions),
 		)
 
 		if err != nil {
 			cmd.Printf("\nError requesting update: %s\n", err)
+			cmd.Println("Should you include the --overwrite flag?")
```
Comment (tony-landreth) [inline internal/cmd/backup.go:125]: I'm not married to the verbiage, but it seemed like a hint might be nice.

File: docs/content/reference/pgo_backup.md [hunk 2]
```diff
@@ -37,6 +37,7 @@ pgo backup CLUSTER_NAME [flags]
 ```
   -h, --help                  help for backup
       --options stringArray   options for taking a backup; can be used multiple times
+      --overwrite             overwrite the backup label
```
Comment (benjaminjb) [inline docs/content/reference/pgo_backup.md:40]:
```suggestion
      --overwrite             overwrite the backup annotation
```

File: internal/cmd/backup.go [hunk 3]
```diff
@@ -58,10 +58,11 @@ pgo backup hippo --repoName="repo1" --options="--type=full"
 	// Limit the number of args, that is, only one cluster name
 	cmdBackup.Args = cobra.ExactArgs(1)
 
-	// `backup` command accepts `repoName` and `options` flags;
+	// `backup` command accepts `repoName`, `overwrite` and `options` flags;
 	// multiple options flags can be used, with each becoming a new line
 	// in the options array on the spec
 	backup := pgBackRestBackup{}
+	cmdBackup.Flags().BoolVar(&backup.Overwrite, "overwrite", false, "overwrite the backup label")
```
Comment (benjaminjb) [inline internal/cmd/backup.go:65]:
```suggestion
	cmdBackup.Flags().BoolVar(&backup.Overwrite, "overwrite", false, "overwrite the backup annotation")
```

File: internal/cmd/backup.go (hunk 1 above)
Comment (benjaminjb) [reply internal/cmd/backup.go:125]: Hmmm, yeah. Maybe something that indicate when they might include that? `If you're getting an ownership conflict, consider the --overwrite flag.`

Eh, might want to think on this a little more.

Comment (benjaminjb) [reply internal/cmd/backup.go:125]: Actually, maybe in the `Long` usage, we should include something about this flag?

Comment (benjaminjb) [reply internal/cmd/backup.go:125]: Maybe put another example in `cmdBackup.Example`?

File: testing/kuttl/e2e/backup/02--backup-with-flags.yaml [hunk 4]
```diff
@@ -2,7 +2,7 @@ apiVersion: kuttl.dev/v1beta1
 kind: TestStep
 commands:
 - script: |
-    RESULT=$(kubectl-pgo --namespace "${NAMESPACE}" backup backup-cluster --repoName repo1 --options="--type=full")
+    RESULT=$(kubectl-pgo --namespace "${NAMESPACE}" backup backup-cluster --repoName repo1 --options="--type=full" --overwrite)
```
Comment (benjaminjb) [inline testing/kuttl/e2e/backup/02--backup-with-flags.yaml:5]: Why this change here?

File: internal/cmd/backup.go (hunk 3 above)
Comment (benjaminjb) [reply internal/cmd/backup.go:65]: Wait, this won't just overwrite the annotation, this will overwrite _anything_ that we send through

File: testing/kuttl/e2e/backup/12--backup-with-overwrite.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+apiVersion: kuttl.dev/v1beta1
+kind: TestStep
+commands:
+- script: |
+    kubectl annotate -n "${NAMESPACE}" postgrescluster backup-cluster postgres-operator.crunchydata.com/pgbackrest-backup="$(date)" --overwrite
+    # The --overwrite flag ensures that ownership of the label passes back to kubectl-pgo.
+    kubectl-pgo --namespace "${NAMESPACE}" backup backup-cluster --overwrite
```
Comment (benjaminjb) [inline testing/kuttl/e2e/backup/12--backup-with-overwrite.yaml:7]: Step 10 changes the ownership of the repoName field, so that a backup will fail with other options.

So this step... what should this step do? We can reset the repoName AND reset the annotation AND check that the result of the `backup` command has the expected result.

(Also, please add changes here to the README.)

**Review by benjaminjb — COMMENTED** [review]
LG, a few questions. Also, think we'll want a parallel change made in restore?

File: testing/kuttl/e2e/backup/02--backup-with-flags.yaml (hunk 4 above)
Comment (tony-landreth) [reply testing/kuttl/e2e/backup/02--backup-with-flags.yaml:5]: I'll revert this one. I was looking at the title of the test and thinking that maybe the idea was to use all the flags you might use when running backup. But that's not what's going on.

File: internal/cmd/backup.go [hunk 6]
```diff
@@ -53,15 +53,19 @@ pgo backup hippo
 # Update the 'backups.pgbackrest.manual.repoName' and 'backups.pgbackrest.manual.options' fields
 # on the 'hippo' postgrescluster and trigger a backup
 pgo backup hippo --repoName="repo1" --options="--type=full"
+
```
Comment (benjaminjb) [inline internal/cmd/backup.go:56]: Not your change, but can you look at line 40 -- I don't like that it uses the word "overwrite" when that's not what we do _unless_ you use the overwrite flag

File: internal/cmd/backup.go (hunk 3 above)
Comment (tony-landreth) [reply internal/cmd/backup.go]: That's what I'd expect:

> 	// Force is going to "force" Apply requests. It means user will
	// re-acquire conflicting fields owned by other people. Force
	// flag must be unset for non-apply patch requests.
	// +optional
	Force *bool `json:"force,omitempty" protobuf:"varint,2,opt,name=force"`

File: docs/content/reference/pgo_backup.md [hunk 7]
```diff
@@ -30,13 +30,17 @@ pgo backup CLUSTER_NAME [flags]
   # Update the 'backups.pgbackrest.manual.repoName' and 'backups.pgbackrest.manual.options' fields
   # on the 'hippo' postgrescluster and trigger a backup
   pgo backup hippo --repoName="repo1" --options="--type=full"
+
```
Comment (benjaminjb) [inline docs/content/reference/pgo_backup.md:33]: Could you change line 12 to... "or by writing those settings using the flags"?

File: internal/cmd/backup.go (hunk 6 above)
Comment (tony-landreth) [reply internal/cmd/backup.go:56]: Does the new text work?

File: crunchy_k8s_support_export_2023-10-24-162950-0400.tar.gz
[the remark this answers is not in the document]
Comment (tony-landreth) [reply crunchy_k8s_support_export_2023-10-24-162950-0400.tar.gz:1]: Thanks for catching that!

**Review by tjmoore4 — APPROVED** [review]

File: internal/cmd/backup.go (hunk 6 above)
Comment (benjaminjb) [reply internal/cmd/backup.go:56]: I think there's ~3 cases here that we (maybe) want to cover:
a) take a backup with the settings already in the spec
b) take a backup with new settings that you write to the spec
c) take a backup with new settings that you write to the spec -- even if someone else owns those settings.

Maybe something like (but better than):
```
Backup allows you to take a backup of a PostgreSQL cluster, either using
the current "spec.backups.pgbackrest.manual" settings on the PostgreSQL cluster
or by writing those settings using the flags (potentially overwriting them with the
`--force-conflicts` flag).
```

(I don't think we actually want parentheses in our docs. Maybe we just want to say "you can take backups with the settings that are there or write those settings" and mention the overwrite situation in the flag description? No, I think people will run into that, we probably want to bring it up in this summary.)

File: internal/cmd/backup.go [hunk 8]
```diff
@@ -53,15 +53,19 @@ pgo backup hippo
...
 
 	// Limit the number of args, that is, only one cluster name
 	cmdBackup.Args = cobra.ExactArgs(1)
 
-	// `backup` command accepts `repoName` and `options` flags;
+	// `backup` command accepts `repoName`, `force-conflicts` and `options` flags;
 	// multiple options flags can be used, with each becoming a new line
 	// in the options array on the spec
 	backup := pgBackRestBackup{}
+	cmdBackup.Flags().BoolVar(&backup.ForceConflicts, "force-conflicts", false, "take ownership and overwrite the backup annotation")
```
Comment (benjaminjb) [inline internal/cmd/backup.go:68]:
```suggestion
	cmdBackup.Flags().BoolVar(&backup.ForceConflicts, "force-conflicts", false, "take ownership and overwrite the backup settings")
```?

File: testing/kuttl/e2e/restore/31--force-conflicts.yaml [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+      - name: repo2
+        volume:
+          volumeClaimSpec:
+            accessModes: [ReadWriteOnce]
+            resources: { requests: { storage: 1Gi } }
+      restore:
+        enabled: true
+        repoName: repo2
+        options:
+        - --db-include=restore-cluster
```
Comment (benjaminjb) [inline testing/kuttl/e2e/restore/31--force-conflicts.yaml:25]: What does the cluster definition do in this file (and in backup/12--backup-with-force)?

Comment (tony-landreth) [reply testing/kuttl/e2e/restore/31--force-conflicts.yaml:25]: YAGNI

File: internal/cmd/backup.go (hunk 6 above)
Comment (tony-landreth) [reply internal/cmd/backup.go:58]: What about this?

File: internal/cmd/restore.go [hunk 10]
```diff
@@ -64,6 +68,8 @@ postgresclusters/hippo patched`)
 	cmd.Flags().StringVar(&restore.RepoName, "repoName", "",
 		"repository to restore from")
 
+	cmd.Flags().BoolVar(&restore.ForceConflicts, "force-conflicts", false, "take ownership and overwrite the restore annotation")
```
Comment (benjaminjb) [inline internal/cmd/restore.go:71]:
```suggestion
	cmd.Flags().BoolVar(&restore.ForceConflicts, "force-conflicts", false, "take ownership and overwrite the restore settings")
```

**Review by benjaminjb — APPROVED** [review]
